core/src/main/kotlin/io/customer/android/core/di/DiGraph.kt

```diff
@@ -1,48 +1,111 @@
 package io.customer.android.core.di
 
 abstract class DiGraph {
+    /**
+     * Map of dependencies that can be overridden with mocks in test functions.
+     */
     val overrides: MutableMap<String, Any> = mutableMapOf()
 
     /**
-     * Call when you want to override a dependency with a mock in test functions.
-     * `di.overrideDependency(CustomerIOInstance::class.java, customerIOMock)`
-     *
-     * See `CustomerIOTest` (or any other class that is top-level of the module) for an example on use.
+     * Map of singleton instances of dependencies in the graph.
+     */
+    val singletons: MutableMap<String, Any> = mutableMapOf()
```

Comment:
why not just use `ConcurrentHashMap` we won't have to use `synchronized` then.

Reply:
same goes for `overrides`

Reply:
This a good idea, but we probably require `computeIfAbsent` to avoid multiple instances here, which was added in [API Level 24](https://developer.android.com/reference/java/util/concurrent/ConcurrentHashMap#computeIfAbsent(K,%20java.util.function.Function%3C?%20super%20K,?%20extends%20V%3E)) unfortunately, while we are on 21 yet. So I have kept `synchronized` block` for now.

For `overrides`, I think we don't need it because we aren't getting and creating instances at same time from it.